### PR TITLE
[SPARK-47079][PYTHON][SQL][CONNECT] Add Variant type info to PySpark

### PR DESCRIPTION
**Pull request by desmondcheongzx**
### What changes were proposed in this pull request?

The Variant datatype was added in https://github.com/apache/spark/pull/43707 but the equivalent PySpark type was not added. In this PR we add Variant to PySpark which allows us to create PySpark dataframes containing the Variant type.

### Why are the changes needed?

Without this PR, trying to create a dataframe containing a variant type results in
`AssertionError: Undefined error message parameter for error class: CANNOT_PARSE_DATATYPE. Parameters: {'error': "Undefined error message parameter for error class: CANNOT_PARSE_DATATYPE. Parameters: {'error': 'variant'}"}`.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Added new PySpark type tests involving Variant.

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

File: python/pyspark/sql/types.py [hunk 1]
```diff
@@ -2384,6 +2408,14 @@ def verify_struct(obj: Any) -> None:
 
         verify_value = verify_struct
 
+    elif isinstance(dataType, VariantType):
+
+        def verify_variant(obj: Any) -> None:
+            # The variant data type can take in any type.
+            pass
+
+        verify_value = verify_variant
```
Comment (HyukjinKwon) [inline python/pyspark/sql/types.py:2417]: Hm, don't we need `createDataFrame`, `collect` and Python UDF test cases that uses this type? If you plan to do them in separate PRs, that's fine.

File: python/pyspark/sql/types.py [hunk 2]
```diff
@@ -1299,6 +1300,12 @@ def fromInternal(self, obj: Tuple) -> "Row":
         return _create_row(self.names, values)
 
 
+class VariantType(AtomicType):
+    """Variant data type, representing semi-structured values."""
```
Comment (xinrong-meng) [inline python/pyspark/sql/types.py:1304]: nit: if it's not for internal use only, shall we add `versionadded`?

**Comment by xinrong-meng** [comment]
Thanks for working on that!

File: python/pyspark/sql/types.py (hunk 1 above)
Comment (desmondcheongzx) [reply python/pyspark/sql/types.py:2417]: Yeah, I was on the fence for adding these tests because we need to add additional python support for variant expressions (i.e. `parse_json` which turns strings into variant data types). Happy to add this support then add these test cases

File: python/pyspark/sql/types.py (hunk 2 above)
Comment (desmondcheongzx) [reply python/pyspark/sql/types.py:1304]: sounds good to me!

File: connector/connect/common/src/main/scala/org/apache/spark/sql/connect/common/DataTypeProtoConverter.scala [hunk 3]
```diff
@@ -270,6 +271,8 @@ object DataTypeProtoConverter {
               .build())
           .build()
 
+      case VariantType => ProtoDataTypes.VariantType
```
Comment (hvanhovell) [inline connector/connect/common/src/main/scala/org/apache/spark/sql/connect/common/DataTypeProtoConverter.scala:274]: This should also have a conversion in the other direction.

Comment (desmondcheongzx) [reply connector/connect/common/src/main/scala/org/apache/spark/sql/connect/common/DataTypeProtoConverter.scala:274]: I might be missing something. Is the conversion above---`case proto.DataType.KindCase.VARIANT => VariantType`---the conversion in the other direction?

File: python/pyspark/sql/types.py [hunk 4]
```diff
@@ -1299,6 +1300,12 @@ def fromInternal(self, obj: Tuple) -> "Row":
         return _create_row(self.names, values)
 
 
+class VariantType(AtomicType):
```
Comment (ueshin) [inline python/pyspark/sql/types.py:1303]: I'm wondering we should make this `AtomicType`? should be just `DataType`?

**Review by HyukjinKwon — APPROVED** [review]

File: python/pyspark/sql/types.py (hunk 4 above)
Comment (desmondcheongzx) [reply python/pyspark/sql/types.py:1303]: (my bad, I thought I'd replied to this last week, but now I see that my comment didn't go through)

In this case I was mirroring the definition of the datatype on the scala side. We can change both in a future PR

**Review by allisonwang-db — APPROVED** [review]

**Review by ueshin — APPROVED** [review]

**Comment by ueshin** [comment]
Thanks! merging to master.

File: connector/connect/common/src/main/scala/org/apache/spark/sql/connect/common/DataTypeProtoConverter.scala (hunk 3 above)
Comment (hvanhovell) [reply connector/connect/common/src/main/scala/org/apache/spark/sql/connect/common/DataTypeProtoConverter.scala:274]: Yeah the conversion to the other side.

Connect can also pass a schema to the server side.
